libsplinter/src/node_registry/unified.rs

```diff
@@ -0,0 +1,359 @@
+
+        let unified = UnifiedNodeRegistry::new(Box::new(writable), vec![Box::new(readable)]);
+
+        let retreived_node = unified.fetch_node("node1").expect("Unable to fetch node");
+
+        assert_eq!(
+            node!("node1", "meta_b" => "b value"),
+            retreived_node
+        );
+    }
+
+    /// This test ensures that the results are unified into a single iterator and nodes are
+    /// replaced with the values from the local store, if the exist.
+    #[test]
+    fn test_unified_iteration_local_selected() {
+        let readable = MemRegistry::default();
+        readable
+            .insert_node(node!("node1", "meta_a" => "a value"))
+            .expect("Unable to insert node");
+        readable
+            .insert_node(node!("node2", "meta_c" => "c value"))
+            .expect("Unable to insert node");
+
+        let writable = MemRegistry::default();
+        writable
+            .insert_node(node!("node1", "meta_b" => "b value"))
+            .expect("Unable to insert node");
+
+        let unified = UnifiedNodeRegistry::new(Box::new(writable), vec![Box::new(readable)]);
+
+        let mut iterator = unified.list_nodes(&[]).expect("Unable to list nodes");
+
+        assert_eq!(
+            Some(node!("node1", "meta_b" => "b value")),
+            iterator.next()
+        );
+        assert_eq!(
+            Some(node!("node2", "meta_c" => "c value")),
+            iterator.next()
+        );
+
+        assert_eq!(None, iterator.next());
+    }
+
+    /// This test ensures that a node which is selected by MetadataPredicates is returned
+    /// regardless of source.
+    ///
+    /// Likewise, nodes that should be filtered out based on properties in one source should not be
+    /// returned in the final iteration.
+    #[test]
+    fn test_unified_iteration_filtering() {
+        let readable = MemRegistry::default();
+        readable
+            .insert_node(node!("node1", "meta_a" => "a value"))
+            .expect("Unable to insert node");
+        readable
+            .insert_node(node!("node2", "meta_c" => "c value"))
+            .expect("Unable to insert node");
+
+        let writable = MemRegistry::default();
+        writable
+            .insert_node(node!("node1", "meta_b" => "b value"))
+            .expect("Unable to insert node");
+
+        let unified = UnifiedNodeRegistry::new(Box::new(writable), vec![Box::new(readable)]);
+
+        let predicates = vec![MetadataPredicate::eq("meta_b", "b value")];
+        let mut iterator = unified
+            .list_nodes(&predicates)
+            .expect("Unable to list nodes");
+
+        assert_eq!(
+            Some(node!("node1", "meta_b" => "b value")),
+            iterator.next()
+        );
+
+        let predicates = vec![MetadataPredicate::ne("meta_b", "b value")];
+        let mut iterator = unified
+            .list_nodes(&predicates)
+            .expect("Unable to list nodes");
+
+        assert_eq!(None, iterator.next());
```

Comment:
Probably not relevant to this PR, but should node2 actually be returned here? Does the ne predicate mean "filter out all nodes that don't have this metadata or have the metadata and it doesn't match" or "filter out all nodes that have this metadata and it doesn't match"? I would assume the latter if I was using this. Maybe we need a "has metadata field" predicate?

Reply:
Added a commit to fix this, and updated the test to cover the scenario.